### PR TITLE
Ping/health analytics

### PR DESCRIPTION
**Pull request by Awesomeman007**
- get health analytics by elderly
- get health analytics by caretaker

### CODE REVIEW

File: src/healthRecord/dto/healthRecord.dto.ts [hunk 1]
```diff
@@ -28,7 +46,40 @@ export class HealthRecordTableDto {
...
+export class HealthRecordAnalyticsDto {
+  @ApiProperty()
+  @IsString()
+  tableName: string;
+
+  @ApiProperty()
+  @IsString()
+  columnName: string;
+
+  @ApiProperty()
```
Comment (tum210544) [inline src/healthRecord/dto/healthRecord.dto.ts:77]: @IsString() is missed?

File: src/healthRecord/healthRecord.controller.ts [hunk 2]
```diff
@@ -24,27 +24,76 @@ export class HealthRecordController {
...
+  @UseGuards(CaretakerGuard)
+  @ApiOkResponse({ type: HealthRecordAnalyticsDto })
+  @ApiNotFoundResponse({ description: 'Elderly Does Not Exist' })
+  @ApiNotFoundResponse({ description: 'Health Column Name Not Found' })
+  @ApiNotFoundResponse({ description: 'Health Record Name Not Found' })
+  @ApiParam({ name: 'elderlyId', type: String, description: 'id of elderly user wants to query' })
+  @ApiParam({ name: 'healthRecordName', type: String, description: 'health record name user wants to find' })
+  @ApiParam({ name: 'columnName', type: String, description: 'column name in health record user wants to query' })
+  @ApiParam({ name: 'timespan', type: String, enum: Timespan, description: 'timespan of data user wants to query' })
+  @Get('analytics/:elderlyId/:healthRecordName/:columnName/:timespan')
```
Comment (tum210544) [inline src/healthRecord/healthRecord.controller.ts:84]: I think every controller need this `@UsePipes(new ValidationPipe({ whitelist: true }))` to apply class validator

File: src/healthRecord/healthRecord.controller.ts [hunk 3]
```diff
@@ -24,27 +24,76 @@ export class HealthRecordController {
...
+  async getHealthAnalyticsByCaretaker(
+    @Param('elderlyId') elderlyId: string,
+    @Param('healthRecordName') healthRecordName: string,
+    @Param('columnName') columnName: string,
+    @Param('timespan') timespan: Timespan = Timespan.Year,
+    @Request() req
+  ): Promise<HealthRecordAnalyticsDto> {
+    const caretakerId = req.user.uid;
+    console.log(elderlyId);
```
Comment (tum210544) [inline src/healthRecord/healthRecord.controller.ts]: delete this

File: src/healthRecord/healthRecord.service.ts [hunk 4]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
+    };
+
+    const currentTimeUnix = new Date().getTime();
+    const millisecInOneDay = 86400000;
+    let startQueryUnix;
+
+    switch (timespan) {
+      case Timespan.Week:
+        startQueryUnix = currentTimeUnix - millisecInOneDay;
+        console.log('in week');
```
Comment (tum210544) [inline src/healthRecord/healthRecord.service.ts]: delete

File: src/healthRecord/healthRecord.service.ts [hunk 5]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
+        break;
+      case Timespan.TwoWeeks:
+        startQueryUnix = currentTimeUnix - 2 * millisecInOneDay;
+        console.log('in 2 week');
```
Comment (tum210544) [inline src/healthRecord/healthRecord.service.ts]: this too

File: src/healthRecord/healthRecord.service.ts [hunk 6]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
+        break;
+      case Timespan.Month:
+        console.log('in month');
```
Comment (tum210544) [inline src/healthRecord/healthRecord.service.ts]: this too

File: src/healthRecord/healthRecord.service.ts [hunk 7]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
+        startQueryUnix = currentTimeUnix - 30 * millisecInOneDay;
+        console.log(startQueryUnix);
```
Comment (tum210544) [inline src/healthRecord/healthRecord.service.ts]: *

File: src/healthRecord/healthRecord.service.ts [hunk 8]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
+        break;
+      case Timespan.ThreeMonths:
+        startQueryUnix = currentTimeUnix - 90 * millisecInOneDay;
+        console.log('in 3 month');
```
Comment (tum210544) [inline src/healthRecord/healthRecord.service.ts]: *

File: src/healthRecord/healthRecord.service.ts [hunk 9]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
+        break;
+      case Timespan.Year:
+        startQueryUnix = currentTimeUnix - 365 * millisecInOneDay;
+        console.log('in year');
```
Comment (tum210544) [inline src/healthRecord/healthRecord.service.ts]: *

File: src/healthRecord/healthRecord.service.ts [hunk 10]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
     const extractedColumns = [...new Map(allColumns.map((column) => [column[distinctiveKey], column])).values()];
```
Comment (sad7898) [inline src/healthRecord/healthRecord.service.ts:145]: How is this different from the `allColumns`? Just curious

File: src/healthRecord/healthRecord.service.ts [hunk 11]
```diff
@@ -90,4 +145,93 @@ export class HealthRecordService {
...
     return extractedColumns;
```
Comment (sad7898) [inline src/healthRecord/healthRecord.service.ts:146]: Don't use `any` type if possible

File: src/healthRecord/healthRecord.service.ts [hunk 12]
```diff
@@ -27,32 +75,39 @@ export class HealthRecordService {
...
     const columnNumbers = columnNames.length;
-    const healthData = [];
+    if (format === 'analytics' && columnNumbers === 1) {
+      const healthData: HealthAnalyticsDataDto[] = [];
+      healthDataRaw.forEach((h) => {
+        healthData.push({
+          dateTime: h.timestamp,
+          value: h.value,
+        });
+      });
```
Comment (sad7898) [inline src/healthRecord/healthRecord.service.ts]: Can be shorten to
```suggestion
      const healthData: HealthAnalyticsDataDto[] = healthDataRaw.map((h) => ({
      dateTime: h.timestamp,
      value: h.value
      })
```

**Review by sad7898 — COMMENTED** [review]
Looks ok but left some questions

File: src/healthRecord/healthRecord.controller.ts (hunk 2 above)
Comment (Awesomeman007) [reply src/healthRecord/healthRecord.controller.ts:84]: For **Get** method , which has no body to pass, I don't find any purpose of using `@UsePipe`. But feel free to correct me of I am wrong.

File: src/healthRecord/healthRecord.service.ts (hunk 10 above)
Comment (Awesomeman007) [reply src/healthRecord/healthRecord.service.ts:145]: It extracts all the unique value sets according to the specified columns

**Review by tum210544 — APPROVED** [review]

**Pull request merged**